src/julia_gc.c

```diff
@@ -448,11 +457,51 @@ void InitMarkFuncBags(UInt type, TNumMarkFuncBags mark_func)
     TabMarkFuncBags[type] = mark_func;
 }
 
+static inline int JMarkGapObjSafe(void * obj)
+{
+    // only traverse objects internally used by GAP
+    void *ty = jl_typeof(obj);
+    if (ty != datatype_mptr && ty != datatype_bag
+        && ty != datatype_largebag && ty != jl_weakref_type)
+        return 0;
+    return jl_gc_mark_queue_obj(JuliaTLS, (jl_value_t *)obj);
+}
+
 static inline int JMark(void * obj)
 {
+#ifdef VALIDATE_MARKING
+    jl_value_t *ty = jl_typeof(obj);
+    if (jl_gc_internal_obj_base_ptr(ty) != ty)
+        abort();
+    if (!jl_typeis(ty, jl_datatype_type))
+        abort();
+#endif
     return jl_gc_mark_queue_obj(JuliaTLS, (jl_value_t *)obj);
 }
 
+void MarkJuliaObjSafe(void * obj)
```

Comment:
Since we need those functions (`MarkJuliaObjSafe` and `MarkJuliaObj`) in JuliaInterface, could you expose them in a header?

And, out of curiosity, is there a reason that the argument type is `void*` and not `jl_value_t*`?

Reply:
Because `void *` is much easier to expose in a header (doesn't force all GAP code to include Julia headers)

Reply:
Makes sense, thank you for the clarification.

Reply:
Yes, we can definitely put that in a header.